File: symforce/examples/bundle_adjustment_in_the_large/bundle_adjustment_in_the_large.py
# ----------------------------------------------------------------------------
# SymForce - Copyright 2022, Skydio, Inc.
# This source code is under the Apache 2.0 license found in the LICENSE file.
# ----------------------------------------------------------------------------

"""
Symbolic factor and codegen for the Bundle-Adjustment-in-the-Large problem
"""

from pathlib import Path

import symforce.symbolic as sf
from symforce import codegen
from symforce.codegen import values_codegen
from symforce.values import Values


def snavely_reprojection_residual(
    cam_T_world: sf.Pose3,
    intrinsics: sf.V3,
    point: sf.V3,
    pixel: sf.V2,
    epsilon: sf.Scalar,
) -> sf.V2:
    """
    Reprojection residual for the camera model used in the Bundle-Adjustment-in-the-Large dataset, a
    polynomial camera with two distortion coefficients, cx == cy == 0, and fx == fy

    See https://grail.cs.washington.edu/projects/bal/ for more information

    Args:
        cam_T_world: The (inverse) pose of the camera
        intrinsics: Camera intrinsics (f, k1, k2)
        point: The world point to be projected
        pixel: The measured pixel in the camera (with (0, 0) == center of image)

    Returns:
        residual: The reprojection residual
    """
    focal_length, k1, k2 = intrinsics

    # Here we're writing the projection ourselves because this isn't a camera model provided by
    # SymForce.  For cameras in `symforce.cam` we could just create a `sf.PosedCamera` and call
    # `camera.pixel_from_global_point` instead, or we could create a subclass of `sf.CameraCal` and
    # do that.
    point_cam = cam_T_world * point

    p = sf.V2(point_cam[:2]) / sf.Max(-point_cam[2], epsilon)

    r = 1 + k1 * p.squared_norm() + k2 * p.squared_norm() ** 2

    pixel_projected = focal_length * r * p

    return pixel_projected - pixel


def generate(output_dir: Path) -> None:
    """
    Generates the snavely_reprojection_factor into C++, as well as a set of Keys to help construct
    the optimization problem in C++, and puts them into `output_dir`.  This is called by
    `symforce/test/symforce_examples_bundle_adjustment_in_the_large_codegen_test.py` to generate the
    contents of the `gen` folder inside this directory.
    """

    # Generate the residual function (see `gen/snavely_reprojection_factor.h`)
    codegen.Codegen.function(snavely_reprojection_residual, codegen.CppConfig()).with_linearization(
        which_args=["cam_T_world", "intrinsics", "point"]
    ).generate_function(output_dir=output_dir, skip_directory_nesting=True)

    # Make a `Values` with variables used in the C++ problem, and generate C++ Keys for them (see
    # `gen/keys.h`)
    values = Values(
        cam_T_world=sf.Pose3(),
        intrinsics=sf.V3(),
        point=sf.V3(),
        pixel=sf.V2(),
        epsilon=sf.Scalar(),
    )

    values_codegen.generate_values_keys(values, output_dir, skip_directory_nesting=True)

File: symforce/examples/bundle_adjustment_in_the_large/bundle_adjustment_in_the_large.cc
/* ----------------------------------------------------------------------------
 * SymForce - Copyright 2022, Skydio, Inc.
 * This source code is under the Apache 2.0 license found in the LICENSE file.
 * ---------------------------------------------------------------------------- */

#include <algorithm>
#include <fstream>
#include <vector>

#include <spdlog/spdlog.h>

#include <sym/pose3.h>
#include <symforce/opt/factor.h>
#include <symforce/opt/optimizer.h>
#include <symforce/opt/values.h>

#include "./gen/keys.h"
#include "./gen/snavely_reprojection_factor.h"

/**
 * Calculate median position
 */
Eigen::Vector3d CalculateMedianPosition(const sym::Valuesd& values, int num_cameras) {
  if (num_cameras == 0) {
    return Eigen::Vector3d::Zero();
  }

  std::vector<Eigen::Vector3d> positions;
  positions.reserve(num_cameras);
  for (int i = 0; i < num_cameras; i++) {
    positions.push_back(values.At<sym::Pose3d>(sym::Keys::CAM_T_WORLD.WithSuper(i)).Position());
  }

  std::nth_element(
      positions.begin(), positions.begin() + positions.size() / 2, positions.end(),
      [](const Eigen::Vector3d& a, const Eigen::Vector3d& b) { return a.norm() < b.norm(); });

  return positions[positions.size() / 2];
}

/**
 * Remove cameras that are far from the median
 * See https://github.com/symforce-org/symforce/issues/387
 */
void FilterCameraOutliers(std::vector<sym::Factord>& factors, sym::Valuesd& values,
                          int& num_cameras, double threshold = 10.0) {
  std::vector<sym::Pose3d> camera_poses;
  std::vector<sym::Factord> filtered_factors;

  Eigen::Vector3d mean = CalculateMedianPosition(values, num_cameras);

  int num_filtered_cameras = 0;
  for (int i = 0; i < num_cameras; i++) {
    sym::Pose3d pose = values.At<sym::Pose3d>(sym::Keys::CAM_T_WORLD.WithSuper(i));
    double distance = (pose.Position() - mean).norm();

    if (distance <= threshold) {
      values.Set(sym::Keys::CAM_T_WORLD.WithSuper(num_filtered_cameras), pose);
      values.Set(sym::Keys::INTRINSICS.WithSuper(num_filtered_cameras),
                 values.At<Eigen::Vector3d>(sym::Keys::INTRINSICS.WithSuper(i)));

      filtered_factors.push_back(factors[i]);

      num_filtered_cameras++;
    }
  }

  num_cameras = num_filtered_cameras;
  factors = std::move(filtered_factors);
}

using namespace sym::Keys;

/**
 * Create a `sym::Factor` for the reprojection residual, attached to the given camera and point
 * variables.  It's also attached to fixed entries in the Values for the pixel measurement and the
 * constant EPSILON.
 */
sym::Factord MakeFactor(int camera, int point, int pixel) {
  return sym::Factord::Hessian(sym::SnavelyReprojectionFactor<double>,
                               /* all_keys = */
                               {
                                   CAM_T_WORLD.WithSuper(camera),
                                   INTRINSICS.WithSuper(camera),
                                   POINT.WithSuper(point),
                                   PIXEL.WithSuper(pixel),
                                   EPSILON,
                               },
                               /* optimized_keys = */
                               {
                                   CAM_T_WORLD.WithSuper(camera),
                                   INTRINSICS.WithSuper(camera),
                                   POINT.WithSuper(point),
                               });
}

/**
 * A struct to represent the problem definition
 */
struct Problem {
  std::vector<sym::Factord> factors;
  sym::Valuesd values;
  int num_cameras;
  int num_points;
  int num_observations;
};

/**
 * Read the problem description from the given path
 *
 * See https://grail.cs.washington.edu/projects/bal/ for file format description
 */
Problem ReadProblem(const std::string& filename) {
  std::ifstream file(filename);

  int num_cameras, num_points, num_observations;
  file >> num_cameras;
  file >> num_points;
  file >> num_observations;

  std::vector<sym::Factord> factors;
  sym::Valuesd values;

  for (int i = 0; i < num_observations; i++) {
    int camera, point;
    file >> camera;
    file >> point;

    double px, py;
    file >> px;
    file >> py;

    factors.push_back(MakeFactor(camera, point, i));
    values.Set(PIXEL.WithSuper(i), Eigen::Vector2d(px, py));
  }

  for (int i = 0; i < num_cameras; i++) {
    double rx, ry, rz, tx, ty, tz, f, k1, k2;
    file >> rx;
    file >> ry;
    file >> rz;
    file >> tx;
    file >> ty;
    file >> tz;
    file >> f;
    file >> k1;
    file >> k2;

    values.Set(CAM_T_WORLD.WithSuper(i),
               sym::Pose3d(sym::Rot3d::FromTangent(Eigen::Vector3d(rx, ry, rz)),
                           Eigen::Vector3d(tx, ty, tz)));
    values.Set(INTRINSICS.WithSuper(i), Eigen::Vector3d(f, k1, k2));
  }

  for (int i = 0; i < num_points; i++) {
    double x, y, z;
    file >> x;
    file >> y;
    file >> z;

    values.Set(POINT.WithSuper(i), Eigen::Vector3d(x, y, z));
  }

  values.Set(EPSILON, sym::kDefaultEpsilond);

  FilterCameraOutliers(factors, values, num_cameras);

  return {std::move(factors), std::move(values), num_cameras, num_points, num_observations};
}

/**
 * Example usage: `bundle_adjustment_in_the_large_example data/problem-21-11315-pre.txt`
 */
int main(int argc, char** argv) {
  SYM_ASSERT_EQ(argc, 2);

  // Read the problem from disk, and create the Values and factors
  const auto problem = ReadProblem(argv[1]);

  // Create a copy of the Values - we'll optimize this one in place
  sym::Valuesd optimized_values = problem.values;

  // Optimize
  auto params = sym::DefaultOptimizerParams();
  params.verbose = true;
  params.lambda_update_type = sym::lambda_update_type_t::DYNAMIC;
  sym::Optimizerd optimizer{params, std::move(problem.factors)};
  const auto stats = optimizer.Optimize(optimized_values);

  spdlog::info("Finished in {} iterations", stats.iterations.size());
}
